src/Control/Monad/Free/Freer.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE FlexibleInstances, GADTs, MultiParamTypeClasses, TypeFamilies, UndecidableInstances #-}
+module Control.Monad.Free.Freer
+( FreerF(..)
+, Freer(..)
+, liftFreerF
+, iter
+, iterA
+, liftF
+, wrap
+) where
+
+import Control.Monad ((>=>))
+import Control.Monad.Free.Class
+import Data.Bifunctor
+import Data.Functor.Classes
+import Data.Functor.Foldable
+
+data FreerF f a b where
+  Pure :: a -> FreerF f a b
+  Free :: (x -> b) -> f x -> FreerF f a b
+
+liftFreerF :: f b -> FreerF f a b
+liftFreerF = Free id
+
+
+newtype Freer f a = Freer { runFreer :: FreerF f a (Freer f a) }
+
+iter :: Functor f => (f a -> a) -> Freer f a -> a
+iter algebra = cata $ \ r -> case r of
+  Pure a -> a
+  Free t r -> algebra (t <$> r)
+
+iterA :: (Functor f, Applicative m) => (f (m a) -> m a) -> Freer f a -> m a
+iterA algebra = cata $ \ r -> case r of
+  Pure a -> pure a
+  Free t r -> algebra (t <$> r)
+
+
+-- Instances
+
+instance Bifunctor (FreerF f) where
+  bimap f g r = case r of
+    Pure a -> Pure (f a)
+    Free t r -> Free (g . t) r
+
+instance Functor (FreerF f a) where
+  fmap = second
+
+instance Functor (Freer f) where
+  fmap f = Freer . bimap f (fmap f) . runFreer
+
+instance Applicative (Freer f) where
+   pure = Freer . Pure
+   Freer g <*> a = case g of
+     Pure f -> fmap f a
+     Free t r -> Freer (Free ((<*> a) . t) r)
+
+instance Monad (Freer f) where
+  return = pure
+  Freer g >>= f = case g of
+    Pure a -> f a
+    Free t r -> Freer (Free (t >=> f) r)
+
+instance MonadFree f (Freer f) where
+  wrap = Freer . Free id
+
+
+instance Foldable f => Foldable (FreerF f a) where
+  foldMap f g = case g of
+    Pure _ -> mempty
+    Free t r -> foldMap (f . t) r
+
+instance Foldable f => Foldable (Freer f) where
+  foldMap f = foldMap (foldMap f) . runFreer
+
+
+type instance Base (Freer f a) = FreerF f a
+
+instance Recursive (Freer f a) where project = runFreer
+instance Corecursive (Freer f a) where embed = Freer
```

Comment:
I also wrote this module so as to avoid orphans, but this one was mostly cribbing from Oleg.